android/java/org/chromium/chrome/browser/vpn/wireguard/WireguardServiceImpl.java

```diff
@@ -124,6 +128,37 @@ private void updateVpnNotification(String notificationText) {
         mNotificationManager.notify(BRAVE_VPN_NOTIFICATION_ID, notification);
     }
 
+    private void recordSessionTimes() {
+        long sessionStartTimeMs = BraveVpnPrefUtils.getSessionStartTimeMs();
+        long sessionEndTimeMs = BraveVpnPrefUtils.getSessionEndTimeMs();
+
+        Calendar sessionEndTimeCal = Calendar.getInstance();
+        sessionEndTimeCal.setTimeInMillis(sessionEndTimeMs);
+        Calendar currDate = Calendar.getInstance();
+
+        long currTimeMs = System.currentTimeMillis();
+        if (sessionStartTimeMs < 0) {
+            BraveVpnPrefUtils.setSessionStartTimeMs(currTimeMs);
+        }
+
+        if (currDate.get(Calendar.YEAR) != sessionEndTimeCal.get(Calendar.YEAR)
+                || currDate.get(Calendar.MONTH) != sessionEndTimeCal.get(Calendar.MONTH)
+                || currDate.get(Calendar.DAY_OF_MONTH)
+                        != sessionEndTimeCal.get(Calendar.DAY_OF_MONTH)) {
+            BraveVpnPrefUtils.setSessionEndTimeMs(currTimeMs);
+        }
+    }
+
+    private void updateRecordSessionTimesTimer() {
+        mVpnStatisticsTimer = new Timer();
+        mVpnStatisticsTimer.schedule(new TimerTask() {
+            @Override
+            public void run() {
+                recordSessionTimes();
+            }
+        }, 0, 60000);
```

Comment:
I agree that would be cleaner, but there's no guarantee that the user will kill their VPN connection regularly (which would cause the Wireguard service to be destroyed).

Reply:
What else stops the timer? Are you relying on app close?

It looks like this timer keeps updating the sessions times forever. I'm wondering what events stop it, thus marking the end of the session.

Reply:
This service runs in the background. When the VPN connection is terminated, the service is destroyed, which will destruct the timer. Added a comment for future reference